### PR TITLE
Authentication Promise or Callback Style APIs

### PR DESCRIPTION
**Pull request by mbalex99**
## What, How & Why?

All http authentication methods are now `callback` or `Promise` styles

```javascript
Realm.Sync.User.login(url, username, password, (err, user) => {
    // normal callback
})

// if callback is not specified a Promise<User> is returned
Realm.Sync.User.login(url, username, password)
    .then(user => {})
    .catch(err => {})

// this is great for async await style of programming
try {
     await Realm.Sync.User.login(url, username, password)
} catch (err)  {

}
```

`Realm.open` is now favored over `Realm.openAsync`. I've marked in the comments that `Realm.openAsync` will be deprecated in the future. 
```javascript
Realm.open(config)
    .then(realm => {})
    .catch(err => {})

// if you like callbacks

Realm.open(config, (err, realm) => {
     // this just like `Realm.openAsync`
})
```

Reference to the issue(s) addressed by this PR: #1182 

## ☑️ ToDos

* [ ] 📝 Changelog entry
* [x] 🚦 Tests
* [x] 📝 Public documentation PR created or is not necessary
* [x] 💥 `Breaking` label has been applied or is not necessary

*If this PR adds or changes public API's:*
* [x] typescript definitions file is updated
* [x] jsdoc files updated
* [ ] Chrome debug API is updated if API is available on React Native

### CODE REVIEW

File: docs/sync.js [hunk 1]
```diff
@@ -170,6 +173,7 @@ class User {
      * @param {string} adminToken - existing admin token
      * @param {string} [server] - authentication server
      * @return {User} - admin user populated with the given token and server
+     * @returns {Promise} returns a Promise if the callback was not supplied
```
Comment (fealebenpae) [inline docs/sync.js]: `Realm.Sync.User.adminUser` is synchronous and doesn't return a Promise, nor take a callback.

File: lib/extensions.js [hunk 2]
```diff
@@ -42,8 +42,8 @@ module.exports = function(realmConstructor) {
 
     //Add async open API
     Object.defineProperties(realmConstructor, getOwnPropertyDescriptors({
-        open(config) {
-            return new Promise((resolve, reject) => {
+        open(config, callback) {
```
Comment (fealebenpae) [inline lib/extensions.js]: Don't add a callback here, leave it as it is.

File: lib/extensions.js [hunk 3]
```diff
@@ -59,9 +59,21 @@ module.exports = function(realmConstructor) {
...
+                    .catch(err => {
+                        callback(err, undefined)
+                    });
+            } else {
+                return promise
+            }
         },
 
         openAsync(config, callback) {
+                console.warn('openAsync is deprecated in favor of Realm.open')
```
Comment (fealebenpae) [inline lib/extensions.js]: Not all of our environments have `console.warn`, I think.

File: lib/index.d.ts [hunk 4]
```diff
@@ -263,10 +263,9 @@ declare namespace Realm.Sync {
         readonly server: string;
         readonly token: string;
         static adminUser(adminToken: string, server?: string): User;
-        static login(server: string, username: string, password: string, callback: (error: any, user: User) => void): void;
-        static loginWithProvider(server: string, provider: string, providerToken: string, callback: (error: any, user: User) => void): void;
-        static register(server: string, username: string, password: string, callback: (error: any, user: User) => void): void;
-        static registerWithProvider(server: string, options: { provider: string, providerToken: string, userInfo: any }, callback: (error: Error | null, user: User | null) => void): void;
+        static login(server: string, username: string, password: string, callback?: (error: any, user: User) => void): void | Promise<Realm.Sync.User>;
```
Comment (fealebenpae) [inline lib/index.d.ts]: Wouldn't it be better to declare these as distinct overloads?

File: lib/user-methods.js [hunk 5]
```diff
@@ -163,29 +170,39 @@ module.exports = {
...
+            }
         },
 
         registerWithProvider(server, options, callback) {
 
             // Compatibility with previous signature: 
             // registerWithProvider(server, provider, providerToken, callback)
             if (arguments.length === 4) {
-                checkTypes(arguments, ['string', 'string', 'string', 'function']);
+                checkTypes(arguments, ['string', 'string', 'string']);
```
Comment (fealebenpae) [inline lib/user-methods.js]: Is this correct, given the condition just above?

**Comment by fealebenpae** [comment]
Please target the `master` branch.

**Comment by kristiandupont** [comment]
This seems to be based off of the 2.0 branch, but targeting the master branch. I think it should be based off of the master branch and target that as well.

**Pull request closed without merging**